invenio_i18n/translation_utilities/collect.py

```diff
@@ -0,0 +1,237 @@
+
+        if translations_by_locale:
+            normalized_name = normalize_package_to_module_name(package_name)
+            translations_by_package[normalized_name] = translations_by_locale
+
+            # Store for merged output
+            for locale, translations in translations_by_locale.items():
+                collected_translations[locale][normalized_name] = translations
+
+    return {
+        "translations_by_package": translations_by_package,
+        "collected_translations": dict(collected_translations),
+        "packagesProcessed": len(packages),
+        "locales": sorted(list(collected_translations.keys())),
+    }
+
+
+def write_translations_to_json(
+    collected_data: dict,
+    output_dir: Path,
+) -> None:
+    """Write collected translations to JSON files.
+
+    :param collected_data: Output from collect_translations()
+    :param output_dir: Where to save the translation files
+    """
+    translations_by_package = collected_data["translations_by_package"]
+    collected_translations = collected_data["collected_translations"]
+
+    for normalized_name, translations_by_locale in translations_by_package.items():
+        package_output = output_dir / normalized_name / "translations.json"
+        write_json_file(package_output, translations_by_locale)
+
+    write_json_file(output_dir / "translations.json", collected_translations)
+
+
+def write_validation_report(
+    validation_summary: dict,
+    output_dir: Path,
+) -> None:
+    """Write validation report to JSON file.
+
+    :param validation_summary: Output from validate_translations()
+    :param output_dir: Where to save the validation report
+    """
+    report_path = output_dir / "validation-report.json"
+    write_json_file(report_path, validation_summary)
+
+
+def validate_translations(
+    packages: list[str],
+) -> dict[
+    str,
+    dict[str, int]
+    | dict[str, dict[str, int | list[dict[str, dict[str, int]]]]]
+    | dict[str, dict[str, int | bool]]
+    | list[dict[str, str | dict[str, list[str]] | dict[str, int]]],
+]:
+    """Validate translations from packages.
+
+    :param packages: List of package names to check like ['invenio-app-rdm']
+    :return: Summary of all issues found
+    """
+    all_validation_reports: list[
+        dict[str, str | dict[str, list[str]] | dict[str, int]]
+    ] = []
+
+    for package_name in packages:
+        validation_reports = get_package_validation_reports(package_name)
+        all_validation_reports.extend(validation_reports)
+
+    summary = _calculate_validation_report(all_validation_reports, packages)
+
+    return summary
+
+
+def _calculate_validation_report(
+    reports: list[dict[str, str | dict[str, list[str]] | dict[str, int]]],
+    packages: list[str],
+) -> dict[
+    str,
+    dict[str, int]
+    | dict[str, dict[str, int | list[dict[str, dict[str, int]]]]]
+    | dict[str, dict[str, int | bool]]
+    | list[dict[str, str | dict[str, list[str]] | dict[str, int]]],
+]:
```

Comment:
The nested dictionary return type hints are quite complex. Would it be clearer to use TypedDict or break this into smaller type definitions?